### ISSUE TITLE
Take care of Phoenix asset setup as part of `mix firmware` task

### ISSUE THREAD

**Issue by vanvoljg** [ElixirSeattle/xebow#58]
When building the firmware, the Phoenix assets need to first be fetched, installed, and deployed. While this is easy enough, `mix firmware` doesn't check or know when the Phoenix assets are missing, and it displays no errors or warnings, so the Live View fails to work correctly.

We should make the `mix firmware` (and `mix firmware.burn`) task ask if a user wants to install the Phoenix deps, similar to how `mix phx.new` does.

Phoenix does it like this: https://github.com/phoenixframework/phoenix/blob/1f0040a513c25a0ffee78931357fde44ac5d0583/installer/lib/mix/tasks/phx.new.ex#L150

Xebow might not need to be this involved. It would be really cool if we could make it more automagic and somehow detect whether the Phoenix assets need to be installed, but it's probably not strictly necessary.

Bare minimum would be a couple sentences in the readme about how to do this.

**Comment by doughsay** [comment]
I'd like to clarify that fetching the npm deps (`npm install --prefix assets`), is not the same thing as compiling the assets `npm run deploy --prefix assets`.  One is the analogous to `mix deps.get` and the other to `mix compile`.  It's obvious that `mix firmware` should also `compile`, but it doesn't also do `mix deps.get`.

`mix phx.new` does the equivalent of `mix deps.get` + `npm install --prefix assets`, but doesn't actually compile anything (including assets).

Also also: `mix compile` I think actually DOES compile the assets, no?  Isn't that what this is for? https://github.com/ElixirSeattle/xebow/pull/56/files#diff-6023be6004fce4718dad3dafb576d258R14

Lastly, what I did personally to get the assets into the firmware was `mix phx.digest` then `mix firmware`...  I'm not sure if that was correct, but it did seem to work for me...

**Comment by vanvoljg** [comment]
I remember running `mix phx.digest` once or twice, but I can't remember why, now. I think I was trying to learn about deployments on Gigalixir, or something.

Arguing semantics for just a second: while I agree that `npm install` is analogous to `mix deps.get`, and that `npm run deploy` is to `mix compile`, the Elixir side of things seems to leave the JavaScript stuff alone almost entirely. `npm install` doesn't fetch Elixir deps, and `mix deps.get` doesn't install Node modules.

I'm pretty sure `mix phx.new`, when told to fetch and install dependencies, does run `mix deps.compile` (L160 from the link I posted), then it runs `npm install --prefix assets`, and then it runs webpack (and then it runs `mix ecto.setup`). In `package.json`, the `deploy` script runs webpack in production mode. It's not the development mode that the generator tells you to use when you create a new Phoenix app, but it's pretty similar.

As far as `mix compile`, I think it only compiles the app and dependencies. From what I could see, it didn't do anything with the assets directory. Also, if `mix compile` installed the js assets, then `mix firmware` (which runs `mix compile`) would have pulled and installed the assets without issue.

TL;DR: There's Elixir dependencies and there's Node dependencies. Elixir doesn't care if Node has been correctly setup and will trundle happily along and create a slightly broken app.

**Comment by amclain** [comment]
>`npm` install doesn't fetch Elixir deps, and `mix deps.get` doesn't install Node modules.

Trust me, this is a good thing. It helps keep the Elixir and JavaScript ecosystems encapsulated.

>TL;DR: There's Elixir dependencies and there's Node dependencies. Elixir doesn't care if Node has been correctly setup and will trundle happily along and create a slightly broken app.

The first part of what I'm going to describe is probably no surprise, but I'm going to progressively expand on the following diagrams. There are two separate software ecosystems we're talking about: Elixir and JavaScript. Each of these are separate languages with their own sets of tooling, package management, dependency versioning, etc. Keeping them separated means one language's decisions don't affect the other.

![elixir](https://user-images.githubusercontent.com/4755047/86883296-92e7f280-c0a6-11ea-959f-82725159f707.png)

![js](https://user-images.githubusercontent.com/4755047/86883301-94b1b600-c0a6-11ea-8dff-211c694df6e4.png)

When you're working on a Phoenix project, that JavaScript ecosystem lives in `/assets`. Otherwise you're working with Elixir. Phoenix has tooling around this already, like a file watcher that recompiles assets when they change and does a live reload in development. They also have a method for compiling the assets along with the Elixir bit for when you want to create a deployment.

![phoenix](https://user-images.githubusercontent.com/4755047/86884206-1524e680-c0a8-11ea-9ee9-dbb28a9cdee2.png)

What we're doing with Xebow is a Nerves project with a Phoenix application that runs inside it. Nerves provides the framework and tooling to get the embedded system up and running and booted into the BEAM VM, and starts our Elixir app(s). Phoenix is like a sidecar application that runs next to the Xebow application on the BEAM. This is a very simplified diagram of what this system starts to look like.

![nerves-phoenix](https://user-images.githubusercontent.com/4755047/86884400-5b7a4580-c0a8-11ea-8c67-7a158e604540.png)

I think the pain point you're facing is less about any one framework suffering, but the fact that things get weird when frameworks like Nerves and Phoenix get combined, especially since each of their project generators set you up differently (for good reason). The good news is that we should be able to solve the UX gap you're facing within our project. We can contribute any good ideas we come up with back to the community.

**Comment by vanvoljg** [comment]
https://github.com/ElixirSeattle/xebow/pull/79
